### PR TITLE
xonsh: fix wrapper, 0.17.0 -> 0.18.2

### LINKED ISSUE #248978: Xonsh adds unwanted entries to PATH
### Describe the bug
Using xonsh in a `nix-shell` environment can override the Python interpreter, which is a problem if the environment is supposed to use a specific one.

### Steps To Reproduce
```bash
sudo docker run -ti --rm nixos/nix bash
nix-channel --update
```
Then create `shell.nix` containing
```nix
{ pkgs ? import <nixpkgs> {} }:

pkgs.mkShell {
  packages = [
    (pkgs.python311.withPackages (ps: [
      ps.requests
    ]))
    pkgs.xonsh
  ];
}
```
Then inside xonsh, a different Python interpreter is used:
```bash
nix-shell
xonsh
python -c "import requests"  # -> ModuleNotFoundError: No module named 'requests'
```

### Expected behavior
Running xonsh should not modify the `PATH` such that a different `python` executable is used. In the above, when we run `python` we should get the `requests` library.

### Workarounds

Running `$PATH.pop(0)` removes the problematic PATH entry but this is pretty hacky.

### Notify maintainers

@vrthra

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.35`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

### LINKED ISSUE #326339: Xonsh extraPacakges doesn't work
### Describe the bug

The wrapper introduced in #240246 is supposed to accept code like: `xonsh.override { extraPackages = ps: [ ps.psutil ] }`, and then `psutil` should be importable in the resulting Xonsh package. However, when building this package, I am met with:

<summary>
<pre><code>
error: function 'anonymous lambda' called with unexpected argument 'extraPackages'
at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/pkgs/by-name/xo/xonsh/package.nix:1:1:
</code></pre>
<details>
<pre><code>
error:
       … while calling a functor (an attribute set with a '__functor' attribute)
         at /nix/store/81ai25ifhr0l8rryf83h4pj8k0q2pjaj-source/flake.nix:4:37:
            3|   outputs = { self, nixpkgs }: {
            4|     packages.x86_64-linux.default = nixpkgs.legacyPackages.x86_64-linux.xonsh.override {
             |                                     ^
            5|       extraPackages = ps: [ ps.psutil ];

       … from call site
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/trivial.nix:955:7:
          954|     { # TODO: Should we add call-time "type" checking like built in?
          955|       __functor = self: f;
             |       ^
          956|       __functionArgs = args;

       … while calling anonymous lambda
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:144:34:
          143|       # Re-call the function but with different arguments
          144|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                  ^
          145|       # Change the result of the function call by applying g to it

       … while calling a functor (an attribute set with a '__functor' attribute)
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:144:43:
          143|       # Re-call the function but with different arguments
          144|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                           ^
          145|       # Change the result of the function call by applying g to it

       … from call site
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/trivial.nix:955:7:
          954|     { # TODO: Should we add call-time "type" checking like built in?
          955|       __functor = self: f;
             |       ^
          956|       __functionArgs = args;

       … while calling anonymous lambda
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:136:17:
          135|     in
          136|     mirrorArgs (origArgs:
             |                 ^
          137|     let

       … while evaluating a branch condition
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:148:7:
          147|     in
          148|       if isAttrs result then
             |       ^
          149|         result // {

       … while calling the 'isAttrs' builtin
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:148:10:
          147|     in
          148|       if isAttrs result then
             |          ^
          149|         result // {

       … from call site
         at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/lib/customisation.nix:138:16:
          137|     let
          138|       result = f origArgs;
             |                ^
          139|

       error: function 'anonymous lambda' called with unexpected argument 'extraPackages'
       at /nix/store/nm7g9fgcffl6nar6vpib220b51dyv0fl-source/pkgs/by-name/xo/xonsh/package.nix:1:1:
            1| {
             | ^
            2|   lib,
</code></pre>
</details>
</summary>

Is there a different way of getting to the wrapped Xonsh?

### Steps To Reproduce

To reproduce:

```
$ cat <<EOF > flake.lockb
{
  "nodes": {
    "nixpkgs": {
      "locked": {
        "lastModified": 1720712503,
        "narHash": "sha256-1u2FWxVZfuoMe8mfRvXBuZq6+DrEqeBT8A4uFOV+my8=",
        "owner": "nixos",
        "repo": "nixpkgs",
        "rev": "69214562aeca0d6b28f8a233748bbd6f32a52a3f",
        "type": "github"
      },
      "original": {
        "owner": "nixos",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "nixpkgs": "nixpkgs"
      }
    }
  },
  "root": "root",
  "version": 7
}
EOF

$ cat <<EOF > flake.nix
{
  inputs.nixpkgs.url = "github:nixos/nixpkgs/";
  outputs = { self, nixpkgs }: {
    packages.x86_64-linux.default = self.packages.x86_64-linux.xonsh.override {
      extraPackages = ps: [ ps.psutil ];
    };
  };
}
EOF

$ nix build
```

### Expected behavior
Xonsh should build.

### Additional context

Possibly relevant issues:
- 2023-07-07 #240246
- 2023-03-08 #294328
- 2023-03-22 #297628
- 2023-04-04 #301444 ("likely caused by #297628")
- 2023-04-05 #301449 (resolves #301444)

### Notify maintainers

@samlukeyes123

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, NixOS, 24.11 (Vicuna), 24.11.20240709.feb2849`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.0pre20240627_b44909ac`
 - nixpkgs: `/home/sam/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by SamLukeYes**
## Description of changes

Reverting the breaking change of top-level xonsh package introduced by https://github.com/NixOS/nixpkgs/pull/314728, and restore the previous wrapper behavior.

Fixes https://github.com/NixOS/nixpkgs/issues/248978
Fixes https://github.com/NixOS/nixpkgs/issues/326339

Note:  The fix makes the `xonsh` executables in the unwrapped variants unusable without a python env, so either add it to a python env or use the wrapper instead. To use another Python version (e.g. avoid the f-string regression https://github.com/xonsh/xonsh/issues/5166), override the `python3` used by top-level `xonsh`, like `pkgs.xonsh.override { python3 = pkgs.python311; }`.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by greg-hellings — APPROVED** [review]
OK, so the basic package name is going back to the wrapper and the unwrapped is coming back. :+1:

File: pkgs/by-name/xo/xonsh/unwrapped.nix [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      homepage = "https://xon.sh/";
+      description = "A Python-ish, BASHwards-compatible shell";
+      changelog = "https://github.com/xonsh/xonsh/raw/main/CHANGELOG.rst";
+      license = with lib.licenses; [ bsd3 ];
+      mainProgram = "xonsh";
+      maintainers = with lib.maintainers; [ samlukeyes123 ];
+    };
+  };
+in
+pythonPackages.buildPythonPackage argset
```
Comment (eclairevoyant) [inline pkgs/by-name/xo/xonsh/unwrapped.nix:124]: why the let-in instead of defining it directly?

File: pkgs/by-name/xo/xonsh/unwrapped.nix [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    passthru = {
+      shellPath = "/bin/xonsh";
+      python = pythonPackages.python; # To the wrapper
+      wrapper = throw "The top-level xonsh package is now wrapped. Use it directly.";
+      updateScript = gitUpdater { };
+    };
+
+    meta = {
+      homepage = "https://xon.sh/";
+      description = "A Python-ish, BASHwards-compatible shell";
```
Comment (eclairevoyant) [inline pkgs/by-name/xo/xonsh/unwrapped.nix:116]:
```suggestion
      description = "Python-ish, BASHwards-compatible shell";
```

File: pkgs/by-name/xo/xonsh/unwrapped.nix (hunk 1 above)
Comment (SamLukeYes) [reply pkgs/by-name/xo/xonsh/unwrapped.nix:124]: Anderson has explained it in https://github.com/NixOS/nixpkgs/pull/305316#discussion_r1572699044, and I didn't change it.

**Comment by SamLukeYes** [comment]
I'm going to update xonsh to 0.18.2 in this PR, because it is not merged for too long, and I don't want to open a separate PR for update and rebase this one. Don't merge before it is done.

**Comment by lf-** [comment]
So I have a question about the new wrapper: how do you make a dev shell that has a custom Python environment that also works when you run `xonsh` inside the dev shell, *and* the dev shell lets you import xonsh?

**Comment by SamLukeYes** [comment]
> So I have a question about the new wrapper: how do you make a dev shell that has a custom Python environment that also works when you run `xonsh` inside the dev shell, _and_ the dev shell lets you import xonsh?

Good question. I didn't think of this use case, but adding the unwrapped xonsh to a python env actually works, because creating a python env effectively wraps it.

```nix
{ pkgs ? import <nixpkgs> {} }:

let
  pythonEnv = pkgs.python3.withPackages (ps: with ps; [
    requests
    xonsh
  ]);
in

pkgs.mkShell {
  packages = [ pythonEnv ];
}
```

This is exactly how the wrapper works, except the wrapper is a proper shell package and only exposes xonsh itself.

**Review by greg-hellings — APPROVED** [review]
This, at last, resolves basically all the outstanding difficulties I've been having with Xonsh. Please merge this ASAP.

**Pull request merged**